### ISSUE TITLE
NuGet package manager using unsupported constructor

### ISSUE THREAD

**Issue by kg** [mono/monodevelop#7155]
The NuGet package management infrastructure appears to be using a constructor that is not available in .NET Standard 2.1, which causes errors in some automated tests on certain newer configurations of mono:

```

2019-02-13T15:44:36.9199870Z Errors and Failures:
2019-02-13T15:44:36.9204890Z 
2019-02-13T15:44:36.9212810Z 1) FSharpNUnitLibraryProject (MonoDevelopTests.Template tests.FSharpNUnitLibraryProject)
2019-02-13T15:44:36.9216000Z    System.AggregateException : One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (An error occurred while retrieving package metadata for 'NUnit.2.6.4' from source 'https://api.nuget.org/v3/index.json'.)))
2019-02-13T15:44:36.9217590Z   ----> System.AggregateException : One or more errors occurred. (One or more errors occurred. (An error occurred while retrieving package metadata for 'NUnit.2.6.4' from source 'https://api.nuget.org/v3/index.json'.))
2019-02-13T15:44:36.9218860Z   ----> System.AggregateException : One or more errors occurred. (An error occurred while retrieving package metadata for 'NUnit.2.6.4' from source 'https://api.nuget.org/v3/index.json'.)
2019-02-13T15:44:36.9220540Z   ----> System.InvalidOperationException : An error occurred while retrieving package metadata for 'NUnit.2.6.4' from source 'https://api.nuget.org/v3/index.json'.
2019-02-13T15:44:36.9221390Z   ----> NuGet.Protocol.Core.Types.FatalProtocolException : An error occurred while retrieving package metadata for 'NUnit.2.6.4' from source 'https://api.nuget.org/v3/index.json'.
2019-02-13T15:44:36.9222190Z   ----> System.MissingMethodException : Method not found: void System.Net.Http.StreamContent..ctor(System.IO.Stream,System.Threading.CancellationToken)
2019-02-13T15:44:36.9222290Z 
2019-02-13T15:44:36.9222850Z --AggregateException
2019-02-13T15:44:36.9223830Z   at MonoDevelop.FSharp.Shared.AsyncTaskBind.continuation@248-1 (System.Threading.Tasks.Task t) [0x0000f] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/external/fsharpbinding/MonoDevelop.FSharp.Shared/Extensions.fs:0 
2019-02-13T15:44:36.9224760Z   at MonoDevelop.FSharp.Shared.AsyncTaskBind+AsyncBuilder-Bind@257.Invoke (System.Threading.Tasks.Task delegateArg0) [0x00000] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/external/fsharpbinding/MonoDevelop.FSharp.Shared/Extensions.fs:257 
2019-02-13T15:44:36.9225770Z   at System.Threading.Tasks.ContinuationResultTaskFromTask`1[TResult].InnerInvoke () [0x00024] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-02/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/TaskContinuation.cs:108 
2019-02-13T15:44:36.9226710Z   at System.Threading.Tasks.Task.Execute () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-02/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:2319 
2019-02-13T15:44:36.9227410Z --AggregateException
2019-02-13T15:44:36.9228340Z   at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00013] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-02/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:2029 
2019-02-13T15:44:36.9229360Z   at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-02/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:2759 
2019-02-13T15:44:36.9230620Z   at System.Threading.Tasks.Task.Wait () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-02/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:2625 
2019-02-13T15:44:36.9231540Z   at MonoDevelop.PackageManagement.InstallNuGetPackageAction.Execute (System.Threading.CancellationToken cancellationToken) [0x0005a] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/InstallNuGetPackageAction.cs:135 
2019-02-13T15:44:36.9231950Z   at MonoDevelop.PackageManagement.InstallNuGetPackageAction.Execute () [0x00000] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/InstallNuGetPackageAction.cs:121 
2019-02-13T15:44:36.9232350Z   at MonoDevelop.PackageManagement.Tests.Helpers.NuGetPackageInstaller+<>c__DisplayClass0_0.<InstallPackages>b__1 () [0x00027] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement.Tests/MonoDevelop.PackageManagement.Tests.Helpers/NuGetPackageInstaller.cs:54 
2019-02-13T15:44:36.9233910Z   at System.Threading.Tasks.Task.InnerInvoke () [0x0000f] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-02/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:2476 
2019-02-13T15:44:36.9235440Z   at System.Threading.Tasks.Task.Execute () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-02/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:2319 
2019-02-13T15:44:36.9237150Z --- End of stack trace from previous location where exception was thrown ---
2019-02-13T15:44:36.9237310Z 
2019-02-13T15:44:36.9237550Z   at MonoDevelop.PackageManagement.Tests.Helpers.NuGetPackageInstaller.InstallPackages (MonoDevelop.Projects.Solution solution, System.Collections.Generic.IList`1[T] packages) [0x000ec] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement.Tests/MonoDevelop.PackageManagement.Tests.Helpers/NuGetPackageInstaller.cs:48 
2019-02-13T15:44:36.9238600Z --InvalidOperationException
2019-02-13T15:44:36.9238950Z   at NuGet.PackageManagement.NuGetPackageManager.PreviewInstallPackageAsync (NuGet.ProjectManagement.NuGetProject nuGetProject, NuGet.Packaging.Core.PackageIdentity packageIdentity, NuGet.PackageManagement.ResolutionContext resolutionContext, NuGet.ProjectManagement.INuGetProjectContext nuGetProjectContext, System.Collections.Generic.IEnumerable`1[T] primarySources, System.Collections.Generic.IEnumerable`1[T] secondarySources, System.Threading.CancellationToken token) [0x00a55] in <747438c4536f4d6cb8a05bba4076e466>:0 
2019-02-13T15:44:36.9239440Z   at MonoDevelop.PackageManagement.InstallNuGetPackageAction.ExecuteAsync (System.Threading.CancellationToken cancellationToken) [0x00119] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/InstallNuGetPackageAction.cs:151 
2019-02-13T15:44:36.9240520Z --FatalProtocolException
2019-02-13T15:44:36.9240750Z   at NuGet.Protocol.DependencyInfoResourceV3.ResolvePackage (NuGet.Packaging.Core.PackageIdentity package, NuGet.Frameworks.NuGetFramework projectFramework, NuGet.Protocol.Core.Types.SourceCacheContext cacheContext, NuGet.Common.ILogger log, System.Threading.CancellationToken token) [0x0012b] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9241150Z   at NuGet.PackageManagement.ResolverGather.GatherPackageFromSourceAsync (System.String packageId, NuGet.Versioning.NuGetVersion version, NuGet.Protocol.Core.Types.DependencyInfoResource resource, NuGet.Frameworks.NuGetFramework targetFramework, System.Boolean ignoreExceptions, System.Threading.CancellationToken token) [0x001a3] in <747438c4536f4d6cb8a05bba4076e466>:0 
2019-02-13T15:44:36.9241830Z   at NuGet.PackageManagement.ResolverGather.GatherPackageAsync (NuGet.PackageManagement.ResolverGather+GatherRequest request, System.Threading.CancellationToken token) [0x0022f] in <747438c4536f4d6cb8a05bba4076e466>:0 
2019-02-13T15:44:36.9242060Z   at NuGet.PackageManagement.ResolverGather+<>c__DisplayClass26_1.<StartWorkerTasks>b__0 () [0x0007d] in <747438c4536f4d6cb8a05bba4076e466>:0 
2019-02-13T15:44:36.9242270Z   at NuGet.PackageManagement.ResolverGather.ProcessResultsAsync () [0x000b8] in <747438c4536f4d6cb8a05bba4076e466>:0 
2019-02-13T15:44:36.9242480Z   at NuGet.PackageManagement.ResolverGather.StartTasksAndProcessWork (System.Threading.CancellationToken token) [0x000ef] in <747438c4536f4d6cb8a05bba4076e466>:0 
2019-02-13T15:44:36.9242710Z   at NuGet.PackageManagement.ResolverGather.GatherAsync (System.Threading.CancellationToken token) [0x00289] in <747438c4536f4d6cb8a05bba4076e466>:0 
2019-02-13T15:44:36.9242940Z   at NuGet.PackageManagement.ResolverGather.GatherAsync (NuGet.PackageManagement.GatherContext context, System.Threading.CancellationToken token) [0x0006b] in <747438c4536f4d6cb8a05bba4076e466>:0 
2019-02-13T15:44:36.9243310Z   at NuGet.PackageManagement.NuGetPackageManager.PreviewInstallPackageAsync (NuGet.ProjectManagement.NuGetProject nuGetProject, NuGet.Packaging.Core.PackageIdentity packageIdentity, NuGet.PackageManagement.ResolutionContext resolutionContext, NuGet.ProjectManagement.INuGetProjectContext nuGetProjectContext, System.Collections.Generic.IEnumerable`1[T] primarySources, System.Collections.Generic.IEnumerable`1[T] secondarySources, System.Threading.CancellationToken token) [0x0052c] in <747438c4536f4d6cb8a05bba4076e466>:0 
2019-02-13T15:44:36.9244750Z --MissingMethodException
2019-02-13T15:44:36.9245050Z   at Foundation.NSUrlSessionHandler+NSUrlSessionHandlerDelegate.DidReceiveResponse (Foundation.NSUrlSession session, Foundation.NSUrlSessionDataTask dataTask, Foundation.NSUrlResponse response, System.Action`1[T] completionHandler) [0x0004e] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/Foundation/NSUrlSessionHandler.cs:304 
2019-02-13T15:44:36.9246140Z --- End of stack trace from previous location where exception was thrown ---
2019-02-13T15:44:36.9246230Z 
2019-02-13T15:44:36.9246390Z   at Foundation.NSUrlSessionHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x0022e] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/Foundation/NSUrlSessionHandler.cs:260 
2019-02-13T15:44:36.9246590Z   at MonoDevelop.MacIntegration.NSUrlSessionCredentialsHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x0003b] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MacPlatform/NSUrlSessionCredentialsHandler.cs:58 
2019-02-13T15:44:36.9246830Z   at MonoDevelop.PackageManagement.MonoDevelopServerWarningLogHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x00045] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/MonoDevelopServerWarningLogHandler.cs:25 
2019-02-13T15:44:36.9247020Z   at NuGet.Protocol.StsAuthenticationHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x000bc] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9247210Z   at NuGet.Protocol.NuGetHttpSourceAuthenticationHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x00081] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/NuGetHttpSourceAuthenticationHandler.cs:91 
2019-02-13T15:44:36.9248650Z   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered (System.Threading.Tasks.Task`1[TResult] sendTask, System.Net.Http.HttpRequestMessage request, System.Threading.CancellationTokenSource cts, System.Boolean disposeCts) [0x000b3] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-02/external/bockbuild/builds/mono-x64/external/corefx/src/System.Net.Http/src/System/Net/Http/HttpClient.cs:523 
2019-02-13T15:44:36.9248910Z   at NuGet.Protocol.TimeoutUtility.StartWithTimeout[T] (System.Func`2[T,TResult] getTask, System.TimeSpan timeout, System.String timeoutMessage, System.Threading.CancellationToken token) [0x001bd] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9249090Z   at NuGet.Protocol.HttpRetryHandler.SendAsync (NuGet.Protocol.HttpRetryHandlerRequest request, NuGet.Common.ILogger log, System.Threading.CancellationToken cancellationToken) [0x004a0] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9249300Z   at NuGet.Protocol.HttpSource.GetThrottledResponse (System.Func`1[TResult] requestFactory, System.TimeSpan requestTimeout, System.TimeSpan downloadTimeout, System.Int32 maxTries, System.Guid sessionId, NuGet.Common.ILogger log, System.Threading.CancellationToken cancellationToken) [0x001d6] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9249490Z   at NuGet.Protocol.HttpSource+<>c__DisplayClass14_0`1[T].<GetAsync>b__0 (System.Threading.CancellationToken lockedToken) [0x00287] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9249630Z   at NuGet.Common.ConcurrencyUtilities.ExecuteWithFileLockedAsync[T] (System.String filePath, System.Func`2[T,TResult] action, System.Threading.CancellationToken token) [0x0020f] in <61e956532bfa4e68bbfc0c8842fb6e3a>:0 
2019-02-13T15:44:36.9250080Z   at NuGet.Protocol.HttpSource.GetAsync[T] (NuGet.Protocol.HttpSourceCachedRequest request, System.Func`2[T,TResult] processAsync, NuGet.Common.ILogger log, System.Threading.CancellationToken token) [0x000e7] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9250310Z   at NuGet.Protocol.RegistrationUtility.LoadRanges (NuGet.Protocol.HttpSource httpSource, System.Uri registrationUri, System.String packageId, NuGet.Versioning.VersionRange range, NuGet.Protocol.Core.Types.SourceCacheContext cacheContext, NuGet.Common.ILogger log, System.Threading.CancellationToken token) [0x000e8] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9250530Z   at NuGet.Protocol.ResolverMetadataClient.GetDependencies (NuGet.Protocol.HttpSource httpClient, System.Uri registrationUri, System.String packageId, NuGet.Versioning.VersionRange range, NuGet.Protocol.Core.Types.SourceCacheContext cacheContext, NuGet.Common.ILogger log, System.Threading.CancellationToken token) [0x00087] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9250790Z   at NuGet.Protocol.ResolverMetadataClient.GetRegistrationInfo (NuGet.Protocol.HttpSource httpClient, System.Uri registrationUri, System.String packageId, NuGet.Versioning.VersionRange range, NuGet.Protocol.Core.Types.SourceCacheContext cacheContext, NuGet.Frameworks.NuGetFramework projectTargetFramework, NuGet.Common.ILogger log, System.Threading.CancellationToken token) [0x000b0] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9251050Z   at NuGet.Protocol.DependencyInfoResourceV3.ResolvePackage (NuGet.Packaging.Core.PackageIdentity package, NuGet.Frameworks.NuGetFramework projectFramework, NuGet.Protocol.Core.Types.SourceCacheContext cacheContext, NuGet.Common.ILogger log, System.Threading.CancellationToken token) [0x000d8] in <0cc41fb8f1d742228e77b6939c0d63b7>:0 
2019-02-13T15:44:36.9251170Z 
```

It sounds like it is intended for that ctor to be gone, so the package management stuff needs to not use it. I'm guessing this can be addressed by using a newer version of the nuget libraries that are used to do package management here?

https://devdiv.visualstudio.com/DevDiv/_releaseProgress?_a=release-environment-logs&releaseId=256966&environmentId=1116110

See https://github.com/mono/mono/issues/12996

**Comment by baulig** [comment]
This was an [internal API](https://github.com/mono/mono/blob/2018-06/mcs/class/System.Net.Http/System.Net.Http/StreamContent.cs#L69) in Mono that got removed when we switched to using the CoreFX code.

**Comment by mrward** [comment]
The error seems to be from Xamarin.Mac's NSUrlSessionHandler.

```
MissingMethodException
  at Foundation.NSUrlSessionHandler+NSUrlSessionHandlerDelegate.DidReceiveResponse (Foundation.NSUrlSession session, Foundation.NSUrlSessionDataTask dataTask, Foundation.NSUrlResponse response, System.Action`1[T] completionHandler) [0x0004e] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/Foundation/NSUrlSessionHandler.cs:304 
```


Looking at the NSUrlSessionHandler there is an NSUrlSessionDataTaskStreamContent class used in DidReceiveResponse and that looks like it uses that internal StreamContent's constructor.

https://github.com/xamarin/xamarin-macios/blob/ad57f87754910d7b5a91ea8ec8648c18fd6d7ac1/src/Foundation/NSUrlSessionHandler.cs#L531

```
class NSUrlSessionDataTaskStreamContent : StreamContent
{
	Action disposed;

	public NSUrlSessionDataTaskStreamContent (NSUrlSessionDataTaskStream source, Action onDisposed, CancellationToken token) : base (source, token)
	{

```

```
System.MissingMethodException : Method not found: void System.Net.Http.StreamContent..ctor(System.IO.Stream,System.Threading.CancellationToken)
```

**Comment by baulig** [comment]
Thanks @mrward.  I'm trying to collect everything in https://github.com/mono/mono/issues/12996 (there's also a temporary work-around PR linked in that issue).

**Comment by mandel-macaque** [comment]
@mrward I'll see if we can remove that from the NSUrlSession handler since it might be faster

**Comment by directhex** [comment]
Does a Xamarin.Mac bug exist for NSUrlSessionHandler?

**Comment by mrward** [comment]
Closing this. A fix was applied and is available in Xamarin.Mac 5.16:
https://github.com/xamarin/xamarin-macios/pull/6277

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.6.0.802 (master)

Pull Request #441 merged by: Lluis Sanchez
Author: mono
Commit: e6a40a053238e6af597648b4b2d948e878f84663 (xamarin/monodevelop)
